### ISSUE TITLE
这整个系统是实时因果的吗？

### ISSUE THREAD

**Issue by zuowanbushiwo** [Qinwen-Hu/dparn#4]
@Qinwen-Hu 
  参考[DPCRN_DNS3](https://github.com/Le-Xiaohuai-speech/DPCRN_DNS3)，可以变成实时的一帧一帧流式推理，这个系统是否可以？能实现一下吗？
谢谢！

**Comment by Qinwen-Hu** [comment]
系统是实时因果的，流式推理的代码暂时还没有。以后有空可能会更新。

**Comment by zuowanbushiwo** [comment]
@Qinwen-Hu  期待更新
现在对流式处理还是一头雾水，不知道怎么修改，也没找到相关的资料。
目前里面Modules.py 里面的 Conv2d BatchNorm2d  ConvTranspose2d 是不是都不是因果的？
怎么看哪些nn.Module 是不是因果的，你有经验和资料吗？能请教一下吗？
谢谢！

**Comment by JaySpeech** [comment]
按照我的理解，这个要做流式处理需要做如下处理： 1.选定chunk size，例如4,8,16     2.设定一个固定时长的端点用来重attention的状态，这个固定时长不能太长，否则越到后面越慢。  3.7GMACS的运算量在嵌入式cpu端下的消耗也是挺大的。  不知道理解是否正确？

**Comment by Qinwen-Hu** [comment]
> 

我之前也没有做过流式处理的，DPCRN是我师兄的工作，他有写过这方面的代码。只要某一帧的处理没有用到未来帧的信息，系统就是因果的。Conv2d这类模块可以根据padding来看是否因果。如果有缓存帧，BatchNorm2d应该确实不是因果的，可以换成layernorm之类的。

**Comment by Qinwen-Hu** [comment]
> @Qinwen-Hu 期待更新 现在对流式处理还是一头雾水，不知道怎么修改，也没找到相关的资料。 目前里面Modules.py 里面的 Conv2d BatchNorm2d ConvTranspose2d 是不是都不是因果的？ 怎么看哪些nn.Module 是不是因果的，你有经验和资料吗？能请教一下吗？ 谢谢！


我之前也没有做过流式处理的，DPCRN是我师兄的工作，他有写过这方面的代码。只要某一帧的处理没有用到未来帧的信息，系统就是因果的。Conv2d这类模块可以根据padding来看是否因果。如果有缓存帧，BatchNorm2d应该确实不是因果的，可以换成layernorm之类的。

**Comment by Qinwen-Hu** [comment]
> 按照我的理解，这个要做流式处理需要做如下处理： 1.选定chunk size，例如4,8,16 2.设定一个固定时长的端点用来重attention的状态，这个固定时长不能太长，否则越到后面越慢。 3.7GMACS的运算量在嵌入式cpu端下的消耗也是挺大的。 不知道理解是否正确？
DPARN是在频域做attention，时域用的是单向的LSTM。3.7G MACS的运算量确实不小，这个也是MHA、特别是dual-path MHA的一个问题，参数量很小、可以并行运算，但是计算量通常并不小。

**Comment by zuowanbushiwo** [comment]
好的，谢谢你解释。期待你流式处理的更新。
